labrad/units.py

```diff
@@ -346,7 +348,7 @@ def __hash__(self):
         of messed up: 1001*m != 1.001*km because the latter is not an
         exact number.  C'est la floating point.
         '''
-        return hash(self.inBaseUnits()._value)
+        return hash(self[self.unit.base_unit])
```

Comment:
Hash functions are used when you use something as the key in a dictionary
or a member in a set.  It is a requirement of the python hash API that if
x==y, then hash(x)==hash(y) (the converse is not required).  Therefore, a
simple hash based on the value and the unit would be incorrect as two
different representations of the same number (i.e., 1000 m vs 1 km) would
have different hash values.  Thus conversion to base (SI) units.  Any two
values that compare identical should convert to the same representation in
base units.



On Tue, Mar 17, 2015 at 2:09 PM, joshmutus <notifications@github.com> wrote:

> In labrad/units.py
> <https://github.com/martinisgroup/pylabrad/pull/75#discussion_r26616568>:
>
> > @@ -346,7 +348,7 @@ def __hash__(self):
> >          of messed up: 1001*m != 1.001*km because the latter is not an
> >          exact number.  C'est la floating point.
> >          '''
> > -        return hash(self.inBaseUnits()._value)
> > +        return hash(self[self.unit.base_unit])
>
> Since I'm trying to get up to speed with this, I'm going to ask a bunch of
> naive questions. Why do we have a hash function at all in this, and why
> does the docstring say something about converting from one SI prefix to
> another. Does hash not mean what I think it means?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/75/files#r26616568>.
>